packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,209 @@
+		}
+		return { snapshot, version };
+	}
+
+	public async fetchSnapshot(
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	) {
+		const { snapshot, version } =
+			this.pendingLocalState === undefined
+				? await this.fetchSnapshotCore(specifiedVersion, supportGetSnapshotApi)
+				: { snapshot: this.pendingLocalState.baseSnapshot, version: undefined };
+		const snapshotTree: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
+			? snapshot.snapshotTree
+			: snapshot;
+		if (this.pendingLocalState) {
+			this.snapshot = {
+				tree: this.pendingLocalState.baseSnapshot,
+				blobs: this.pendingLocalState.snapshotBlobs,
+			};
+		} else {
+			assert(snapshotTree !== undefined, "Snapshot should exist");
+			// non-interactive clients will not have any pending state we want to save
+			if (this.offlineLoadEnabled && this.isInteractiveClient) {
+				const blobs = await getBlobContentsFromTree(snapshotTree, this.storageAdapter);
+				this.snapshot = { tree: snapshotTree, blobs };
+			}
+		}
+		return { snapshotTree, version };
+	}
+
+	private async fetchSnapshotCore(
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	): Promise<{ snapshot?: ISnapshot | ISnapshotTree; version?: IVersion }> {
+		if (
+			this.mc.config.getBoolean("Fluid.Container.UseLoadingGroupIdForSnapshotFetch") ===
+				true &&
+			supportGetSnapshotApi === true
+		) {
+			const snapshot = await this.storageAdapter.getSnapshot({
+				versionId: specifiedVersion,
+			});
+			const version: IVersion = {
+				id: snapshot.snapshotTree.id ?? "",
+				treeId: snapshot.snapshotTree.id ?? "",
+			};
+
+			if (snapshot === undefined && specifiedVersion !== undefined) {
+				this.mc.logger.sendErrorEvent({
+					eventName: "getSnapshotTreeFailed",
+					id: version.id,
+				});
+			}
+			return { snapshot, version };
+		}
+		return this.fetchSnapshotTree(specifiedVersion);
+	}
+
+	public setSnapshot(
+		snapshot:
+			| {
+					tree: ISnapshotTree;
+					blobs: ISerializableBlobContents;
+			  }
+			| undefined,
+	) {
+		this.snapshot = snapshot;
+	}
+
+	public async getPendingLocalStateCore(props: IGetPendingLocalStateProps) {
+		return PerformanceEvent.timedExecAsync(
+			this.mc.logger,
+			{
+				eventName: "getPendingLocalState",
+				notifyImminentClosure: props.notifyImminentClosure,
+				savedOpsSize: this.getSavedOps().length,
+				clientId: this.getClientId(),
+			},
+			async () => {
+				if (!this.offlineLoadEnabled) {
+					throw new UsageError(
+						"Can't get pending local state unless offline load is enabled",
+					);
+				}
+				const resolvedUrl = this.getResolvedUrl();
+				assert(
+					resolvedUrl !== undefined && resolvedUrl.type === "fluid",
+					"resolved url should be valid Fluid url",
+				);
+				assert(this.snapshot !== undefined, "no base data");
+				const pendingRuntimeState = await this.getRuntime()?.getPendingLocalState(props);
```

Comment:
since only this function needs the runtime i would just take it in this function, rather than in the constructor
```suggestion
				const pendingRuntimeState = await runtime.getPendingLocalState(props);
```

Reply:
probably do the same with client id

Reply:
in general, we should avoid members unless they are frequently used. this will also make unit testing easier

Reply:
it will also let you get ride of asserts, and push the need to validate to the caller